### ISSUE TITLE
Pd crashes on Windows 10

### ISSUE THREAD

**Issue by nyponen** [pure-data/pure-data#2196]
Pd 0.54.1 sometimes bombs when I update audio settings, usually without any error messages. One time, I got these messages:

```
(Tcl) INVALID COMMAND NAME: invalid command name ".xa65ede70.popup"
    while executing
"$menu entrycget $entry -state"
    (procedure "tk::PostOverPoint" line 32)
    invoked from within
"tk::PostOverPoint $menu $x $y $entry"
    (procedure "tk_popup" line 6)
    invoked from within
"tk_popup ${popup} ${xpopup} ${ypopup} 0"
    (procedure "pdtk_canvas_popup" line 25)
    invoked from within
"pdtk_canvas_popup .xa65ede70 1805 972 1 0 "
    ("uplevel" body line 4303)
    invoked from within
"uplevel #0 $docmds"
```

**Comment by nyponen** [comment]
This occurred on Windows 10.

**Comment by Lucarda** [comment]
Hi,

can you describe what changes you do in **audio settings**?

like: "i change delay from 10 to 20 and pd crashes".

it would be good if we can somehow reproduce the crash in our machines.

**Comment by nyponen** [comment]
I was trying different settings of block size, delay, and sample rate to try to reduce the amount of crackling noise I was getting in a patch with 240 bangs in the main patch, and 18 [osc~] objects and 1 [dac~] object in the subpatch. I was mostly changing the block size, because that's the setting that seemed to help the most. 

This problem is inconsistent. Most of the time, I can change audio settings without a problem. Sometimes it bombs with no messages. I only got these messages once, although I got a different error message on another occasion. Unfortunately, I didn't copy that other error message and I don't remember exactly what it said.

**Comment by Lucarda** [comment]
ok, 

first,

can you change the title of the issue to something more descriptive like: "Windows: random crashes while changing audio settings"  ?


then,

Pd on Windows has 3 audio back-ends: MMIO, ASIO and WASAPI. 

- MMIO is not really recommended, it works as-is for basic stuff and you are not getting anything better by changing its somehow defaults **delay**:`80`ms **block**:`64`. 

- About WASAPI I can't tell much as it's rather new and I had never used on Pd or other audio software (possible the same "safety measures" below apply for WASAPI).

- ASIO: this is the most used back-end in Windows audio software.

When using ASIO as a back-end when you change the **block** you are changing the (sound-card's) ASIO driver settings and **not** Pd's block size.

so when your patch starts to crackle most likely you need to increase the **delay**:`x`ms. Of course you can change the **block** also but keep in mind that this is the driver size for the connection. It's the interaction of these 2 settings that keeps your patch from not crackling. (sometimes you aim for the lowest possible latency setting, sometimes you don't care if latency is very high).

also: to avoid crashes it's recommended that you switch dsp off, change audio settings, and then switch dsp on. (it avoids most crashes but not all as the user can ask for bad settings).

Now that you have read all this try changing the settings in a more safety manner. 



PS: I'm not sure if Pd-devs can warranty no crashes from users fiddling with audio settings.

**Comment by nyponen** [comment]
I use many audio applications on my PC. Pd is the only one that has a noise problem. I've tried MMIO, WASAPI, and ASIO with Pd. 

With MMIO, block size 2048 and delay 80 produced much less noise than block size 64 and delay 40.

WASAPI produces less noise than MMIO when using Pd's default settings for block size and delay, but is still not noise free. Changing block size didn't seem to make a difference with WASAPI. On the other hand, increasing  delay to 80 did eliminate most of the noise.

Pd shows two ASIO options in its menu. One of them only works with the tiny internal speaker in my PC. When I selected the other one trying to get my external monitor speakers, I get this error message from Pd "error opening audio: Illegal combination of I/O devices". So I can't use ASIO.

I have not found a combination of Pd settings that is noise-free. Is there one?

If Pd requires DSP to be off when a user changes audio settings, then Pd should turn the DSP off when the user changes a setting, then turn it back on when the user is done.

**Comment by Spacechild1** [comment]
>  Pd is the only one that has a noise problem.

I guess with "noise" you mean dropouts?

> On the other hand, increasing delay to 80 did eliminate most of the noise.

I don't know what kind of PC you have, but with WASAPI I can get a clean test tone with a block size of 64 samples and a delay of 12 ms.

> Pd shows two ASIO options in its menu. One of them only works with the tiny internal speaker in my PC. When I selected the other one trying to get my external monitor speakers, I get this error message from Pd "error opening audio: Illegal combination of I/O devices". So I can't use ASIO.

Can you post a screenshot of your ASIO input and output devices?

> If Pd requires DSP to be off when a user changes audio settings

Contrary to what @Lucarda said, Pd does *not* require you turn off DSP.

> then Pd should turn the DSP off when the user changes a setting, then turn it back on when the user is done.

This is what Pd already does internally :)

**Comment by Lucarda** [comment]
> Pd shows two ASIO options in its menu. One of them only works with the tiny internal speaker in my PC. When I selected the other one trying to get my external monitor speakers, I get this error message from Pd "error opening audio: Illegal combination of I/O devices". So I can't use ASIO.

@nyponen the error message is quite brief but you get the idea that you must use the same driver in the input **and** in the output. Please check again with correct settings.

**Comment by nyponen** [comment]
> @nyponen the error message is quite brief but you get the idea that you must use the same driver in the input **and** in the output. Please check again with correct settings.

I am using the same driver in both the input and the output and got that message.

Just now I got these error messages from Pd while using WASAPI with delay=60 and block size=64. 
"error -9988: Invalid stream pointer
trying to reopen audio device
successfully reopened audio device"

**Comment by nyponen** [comment]
> I guess with "noise" you mean dropouts?

No, I mean a crackling noise. It is inconsistent. I can play the same sound five times and it might be noise free four times and noisy once. It is related to the number of bangs in the main patch, which doesn't make sense to me. The main patch has a number of bangs. Each time a bang is clicked, it sends three messages to the subpatch which plays the sound. The subpatch has 18 [osc~] objects which feed into one [dac~] object. When I have just a few bangs, I never get any noise. When I have many bangs (about 250) I have noise problems using the same subpatch. The fewer the bangs, the fewer times I get noise.

**Comment by Lucarda** [comment]
> It appears that my sound card doesn't support ASIO, which is why it didn't work for me.

very strange. as @Spacechild1 suggested: can you post a screen-shot of your ASIO input and output config?

> The fewer the bangs, the fewer times I get the crackling noise.

are you sure you are not clipping the output? can you share with us your patch? (just zip it and drag and drop it to the your next post.)

**Comment by umlaeute** [comment]
> `"pdtk_canvas_popup .xa65ede70 1805 972 1 0 "`

this is the start of the backtrace you get.
now, the `pdtk_canvas_popup` is *never* triggered by the audio dialog, it *occurs* only if you right-click on a patch. at least AFAIK (and i'm pretty sure IK :-))

so, i do not think that the problem is directly related to changing the audio setting at all, but you are doing something else (possibly unconsciously)

**Comment by nyponen** [comment]
> so, i do not think that the problem is directly related to changing the audio setting at all, but you are doing something else (possibly unconsciously)

That's quite possible. I was trying a lot of things to get the noise to stop.

**Comment by umlaeute** [comment]
> Another question is why

Please let's not throw all unrelated stuff into this issue (which is about some TCL error you get).

Feel free to open a new ticket if the number of gui objects is making your life cumbersome

**Comment by Spacechild1** [comment]
> I'm using the Sound Blaster Play! 3. According to a comment from Sound Blaster on this page, it does not support ASIO. https://www.amazon.com/ask/questions/Tx1TOGFPMWLYQ1K/ref=ask_ql_ql_al_hza

You can use a generic ASIO driver, though, e.g. ASIO4ALL (https://asio4all.org/)

**Comment by nyponen** [comment]
> > That's quite possible. I was trying a lot of things to get the noise to stop.
> 
> Well... Triggering a right click on a patch seems to be pretty unrelated to "making the noise stop" in the first place. So we (well: you) really have to get an idea what is actually going on in the first place...

The error messages tell you that. Pd crashed. It should be fixed regardless of whether it was related to noise or not.

**Comment by nyponen** [comment]
> 
> You can use a generic ASIO driver, though, e.g. ASIO4ALL (https://asio4all.org/)
> 

I was trying ASIO4ALL when I had had the problem with ASIO that I mentioned earlier.

**Comment by umlaeute** [comment]
i was going to comment that the title "Pd crashes on Windows 10" is way too broad, and has nothing to do with the Tcl/Tk error (that you added in the original post) which - while annoying - is not a *crash* (where the program just dies).

on re-reading, I see that you "*sometimes* get seemingly random crashes, and *only once* you've seen the Tcl/Tk error.
I conclude that the two are separate issues. since the mentioning of the crash takes one (1!) line in the original report (well two (2), if you include the title), and the Tcl/Tk error takes fifteen (15) lines, I see why I was mistaken.
you probably want to edit your original post and remove the Tcl/Tk error altogether.

**Comment by umlaeute** [comment]
> i was going to comment that the title "Pd crashes on Windows 10" is way too broad

which btw it really is.

since you have changed the title from the more specific "crashes when changing audio settings" to the very broad "crashes", I guess you have somehow concluded that "changing the audio settings" is *not* the culprit.

which is a bummer, as i do not see how this would be actionable upon.
there are a number of people who are using Pd on Win10, so it seems to work for them.

if we want to fix this, we really need an idea on the circumstances when Pd crashes.
otherwise i guess we just have to close this issue.

**Comment by Spacechild1** [comment]
> I was trying ASIO4ALL when I had had the problem with ASIO that I mentioned earlier.

ASIO4ALL really works with *any* soundcard. I'm pretty sure you just haven't configured it correctly (which sometimes can be a bit tricky indeed). However, this issue tracker is not really the place for audio driver trouble shooting. If you have more questions about ASIO4ALL or ASIO in general, you can ask on the mailing list, Discourse channel, FB group, etc.

**Comment by umlaeute** [comment]
can we close this issue (it has become a little bit of a kitchen-sink, and thus makes it hard to act upon)?

so far i see three separate problems:
- poor audio quality with your soundcard
- random crashes when changing the audio settings
- a Tcl/Tk error

these issues appear to be unrelated (although you experience them together, as you are trying to change the audio settings to fix the poor quality problem).

so i suggest:
- if you have trouble getting your soundcard to work properly, follow https://github.com/pure-data/pure-data/issues/2196#issuecomment-1959372654 (and if that fails, create a new issue as a last resort)
- if you still get random crashes let's create a new issue (that *only* deals with the crash)
- if you get random Tcl/Tk errors

when creating a new issue, please try to provide step-by-step instructions to reproduce the problem. if a patch is involved, please share the patch.

i'm going to close this in a few days, if nobody objects (or if somebody beats me to closing it)
